test/util: expect_msg_seq()

justinmk opened the pull request:
job_spec.lua on AppVeyor (Windows) often fails like this:

      FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 72: jobs changes to given `cwd` directory
    C:/projects/neovim/test/functional\core\job_spec.lua:81: Expected objects to be the same.
    Passed in:
    (table) {
      [1] = 'notification'
      [2] = 'stdout'
     *[3] = {
        [1] = 0
       *[2] = {
          [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimmSjq1S\0' } } }
    Expected:
    (table) {
      [1] = 'notification'
      [2] = 'stdout'
     *[3] = {
        [1] = 0
       *[2] = {
          [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimmSjq1S\0'
         *[2] = '' } } }
    stack traceback:

`expect_msg_seq()` accepts multiple sequences and will call `next_message()` enough times to compare the shortest sequence. If longer sequences are tried, it will call `next_message()` again as needed.

```
+    expect_msg_seq(
+      { {'notification', 'stdout', {0, {dir, ''} } },
+        {'notification', 'stdout', {0, {''} } },
+        {'notification', 'exit', {0, 0} }
+      },
+      -- Alternative sequence:
+      { {'notification', 'stdout', {0, {dir} } },
+        {'notification', 'stdout', {0, {'', ''} } },
+        {'notification', 'stdout', {0, {''} } },
+        {'notification', 'exit', {0, 0} }
+      }
+    )
```

justinmk committed:
test/util: expect_msg_seq()

job_spec.lua on AppVeyor (Windows) often fails like this:

      FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 72: jobs changes to given `cwd` directory
    C:/projects/neovim/test/functional\core\job_spec.lua:81: Expected objects to be the same.
    Passed in:
    (table) {
      [1] = 'notification'
      [2] = 'stdout'
     *[3] = {
        [1] = 0
       *[2] = {
          [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimmSjq1S\0' } } }
    Expected:
    (table) {
      [1] = 'notification'
      [2] = 'stdout'
     *[3] = {
        [1] = 0
       *[2] = {
          [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimmSjq1S\0'
         *[2] = '' } } }
    stack traceback:

Message chunking is non-deterministic, so we need to try different
variants.

bfredl commented on test/functional/core/job_spec.lua:
```diff
@@ -78,8 +80,17 @@ describe('jobs', function()
+    eq_any({'notification', 'stdout', {0, {dir}}},
+           {'notification', 'stdout', {0, {dir, ''}}},
+           next_msg())
+    eq_any({'notification', 'stdout', {0, {}}},
```
This should never happen (empty list)

justinmk replied:
Why does`{dir}` happen? Do we actually expect a empty line to be sent by libuv? Are we not receiving EOF?

bfredl replied:
"lines" doesn't exist as a fundamental concept, from libuv we receive bytes and only bytes. And we unambiguously represent EOF as `{''}`, empty list never happens.

justinmk replied:
So if `{dir}` is received we should expect the next message to be `{''}` (EOF) , right?

bfredl replied:
Not necessarily. If the issue rather is buffering, then in that case we would expect `{'', ''}` and then `{''}`. I think I wrote a helper function exactly for this situation in another test.

justinmk replied:
> I think I wrote a helper function exactly for this situation in another test.

expect_twostreams ?

bfredl replied:
No, expect_twolines, but  I saw it was to limited, it can't handle a final `\n`, only `\n` between two chunks.  Ideally we should join both functions to a generic `expect_streams` that supports _any_ division into chunks as they semantically are equal.

janlazo commented:
This case consistently happens with the 32-bit build. I don't recall this happening with the 64-bit build.

justinmk commented:
Latest commit is just a test, I am working on a new function that will cover what @bfredl hinted at.

justinmk commented on test/helpers.lua:
```diff
@@ -7,13 +7,33 @@ local check_logs_useless_lines = {
+-- Tries multiple accepted ("expected") values until one succeeds.
+-- First N-1 arguments are the accepted values.
+-- Last (Nth) argument is the "actual" value to be compared.
+local function eq_any(...)
```
Don't need this anymore, I'll probably remove it .

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -308,8 +317,15 @@ describe('jobs', function()
+      -- Alternative sequence:
+      { {'notification', 'stdout', {5, {'foo'} } },
+        {'notification', 'stdout', {5, {'', ''} } }
+      }
```
`expect_msg_seq()` accepts multiple sequences and will call `next_message()` enough times to compare the shortest sequence. If longer sequences are tried, it will call `next_message()` again as needed.

justinmk commented:
Looks like the win32 build receives an extra newline.

```
Passed in:
(table) {
  [1] = {
    [1] = 'notification'
    [2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = {
        [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimHAvNYs\0' } } }
  [2] = {
    [1] = 'notification'
    [2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = {
        [1] = ''
        [2] = '' } } }
 *[3] = {
    [1] = 'notification'
   *[2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = {
        [1] = '' } } } }
Expected:
(table) {
  [1] = {
    [1] = 'notification'
    [2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = {
        [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimHAvNYs\0' } } }
  [2] = {
    [1] = 'notification'
    [2] = 'stdout'
    [3] = {
      [1] = 0
      [2] = {
        [1] = ''
        [2] = '' } } }
 *[3] = {
    [1] = 'notification'
   *[2] = 'exit'
    [3] = {
      [1] = 0
      [2] = 0 } } }
stack traceback:
```

justinmk committed:
test/util: remove eq_any()

It was added in the parent commit, but ended up not being used. And
I can't think of a case where it will be used: instead we would probably
want to generalize expect_msg_seq() if necessary.

bfredl commented:
What i had in mind was actually a function that only took a single array of the output ( I e `{dir, ''}`) or two for interleaved streams, and accepted any division into flushed chunks, i e by accepting any prefix of the current remaining expected output. (and probably also flag for EOF expected or not). But this PR will stop the bleeding for now, I can look into the generalization later.
